File: Blind 75/07 - Trees/230-kth-smallest-element-in-a-bst.py
'''
Given the root of a binary search tree, and an integer k, return the kth smallest value (1-indexed) of all the values of the nodes in the tree.
'''
from time import time
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        res, kth = [0], [0]

        def dfs(node):
            if not node:
                return

            dfs(node.left)
            kth[0] += 1
            if kth[0] == k:
                res[0] = node.val
                return
            dfs(node.right)

        dfs(root)
        return res[0]

    def reference(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        curr = root

        while stack or curr:
            while curr:
                stack.append(curr)
                curr = curr.left
            curr = stack.pop()
            k -= 1
            if k == 0:
                return curr.val
            curr = curr.right

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.kthSmallest(*case))
                else:
                    self.kthSmallest(*case)
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(*case))
                else:
                    self.reference(*case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        (TreeNode(3, TreeNode(1, None, TreeNode(2)), TreeNode(4)), 1),
        (
            TreeNode(
                5,
                TreeNode(3, TreeNode(2, TreeNode(1)), TreeNode(4)),
                TreeNode(6),
            ),
            3,
        ),
    ]
    test.quantify(test_cases)

File: Blind 75/07 - Trees/JavaScript/230-kth-smallest-element-in-a-bst.js
const { performance } = require("perf_hooks");

class TreeNode {
  constructor(val, left, right) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

class Solution {
  method(root, k) {
    let res = null;

    const dfs = (node, kthSmallest) => {
      if (!node) return;

      dfs(node.left, kthSmallest);
      kthSmallest[0] += 1;
      if (kthSmallest[0] === k) {
        res = node.val;
        return;
      }
      dfs(node.right, kthSmallest);
    };

    dfs(root, [0]);

    return res;
  }

  reference(root, k, stack = []) {
    while (k--) {
      root = this.moveLeft(root, stack);

      const isSmallest = k === 0;
      if (isSmallest) return root.val;

      root = root.right;
    }
  }

  moveLeft = (root, stack) => {
    while (root !== null) {
      stack.push(root);
      root = root.left;
    }

    return stack.pop();
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input[0], input[1]));
        else this.method(input[0], input[1]);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input[0], input[1]));
        else this.reference(input[0], input[1]);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [new TreeNode(3, new TreeNode(1, null, new TreeNode(2)), new TreeNode(4)), 1],
  [
    new TreeNode(
      5,
      new TreeNode(3, new TreeNode(2, new TreeNode(1)), new TreeNode(4)),
      new TreeNode(6)
    ),
    3,
  ],
];
test.quantify(testCases);
